Improvement/1st improvement design system

Cuervino opened the pull request:
**Component**: storybook, design system, tests

**Description**:

A big PR but not that difficult to review. On the 6 commits only 2 are really important :
- `storybook: update storybook settings`
- `mdx: add first design system bricks`

You also might want to review it by launching the storybook.
You will need to `npm install` before `npm run storybook` because we added some changes to handle Windows environnement.

The rests are cosmetics changes and fix of tests.

bert-e commented:
# Hello cuervino,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/core-ui/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @carlito-scality

* @nicolas2bert

* @rachedbenmustapha

* @monpote

* @chengyanjin

MonPote committed:
package: add cross-env and update tests settings

Co-authored-by: Cuervino <valamir.miezelis@scality.com>

MonPote committed:
config: update gitignore and prettierrc

Co-authored-by: Cuervino <valamir.miezelis@scality.com>

MonPote committed:
storybook: update storybook settings

- Stories are now require to finish by .stories.js
- Prepare the sorting of the stories

Co-authored-by: Cuervino <valamir.miezelis@scality.com>

MonPote committed:
stories: rename stories to *.stories.js

We add ".stories" ending at the end of stories files because
with MDX, we might want to render custom components that are not stories.

Currently, anything that is exported in a file in /stories/* will be
displayed in storybook.
We want a way to discriminate files in the stories/ folder that will not
be displayed by storybook.
So, I change the `main.js` (previous commit) in order to filtrate this kind of files.

Now you will be able to create custom component to render in your MDX files by creating
files that do not end with `.stories.js`.

MonPote committed:
mdx: add first design system bricks

1) New hierarchy
2) New pages that are not stories
3) Icon list
4) In mdx file, better to import id stories (defined in the related js) rather than creating one in the mdx
5) WIP for chart guideline

Co-authored-by: Cuervino <valamir.miezelis@scality.com>

MonPote committed:
snapshot: improve snapshot testing

    https://github.com/storybookjs/storybook/tree/master/addons/storyshots/storyshots-core#multisnapshotwithoptionsoptions
    Please note that I haven't include the integrityOptions because I don't really know what it does
    and it break the test for no special reason IMO.
    If you find out, what's wrong please create an issue to fix it.

nicolas2bert commented on .storybook/preview.js:
```diff
@@ -32,3 +32,30 @@ const withThemeProvider = (Story, context) => {
+        ["Color", "Icons"],
+        "Components",
+        [
+          "Navigation",
```
Should it be ordered alphabetically for better readability?

Cuervino replied:
For headers on level 1 (`STYLE`, `COMPONENT`), the order should be predefined.
For the Intro and Style section, the order of levels 2 should be predefined.
For the library itself (inside `COMPONENTS`), it could be alphabetical order, however it makes sense to put `Navigation` at the beginning, to set `Input` and Selector next to each other, and also `Chart` and `Progress & loading` aside, as they are rather visual components.
At the end/bottom, it could be miscellaneous components.

![image](https://user-images.githubusercontent.com/28654802/109862089-4f0eb280-7c60-11eb-92ce-5d8aaa787ce2.png)

nicolas2bert replied:
Thank you for this explanation.

nicolas2bert commented on package.json:
```diff
@@ -8,7 +8,7 @@
   "scripts": {
     "analyze": "source-map-explorer 'dist/*.js'",
-    "build": "rimraf dist && NODE_ENV=production babel src/lib --out-dir dist --copy-files && rimraf dist/**/*.test.js && rimraf dist/**/__snapshots__",
+    "build": "rimraf dist && cross-env NODE_ENV=production babel src/lib --out-dir dist --copy-files && rimraf dist/**/*.test.js && rimraf dist/**/__snapshots__",
```
What about using `envname` option to avoid having another dependency? https://babeljs.io/docs/en/options#envname

MonPote replied:
We can, but I don't think it's an issue. `cross-env` is a widely use stable dependency to handle this use case.
And because it's only use during the build, it doesn't have an impact on the size on the build itself.

nicolas2bert replied:
A big reason why you use a library in the first place is to get the benefits of the code without having to worry about all the details. The drawback is that you’ve completely tied your code to these dependencies that you don’t own and don’t control. Instead, you could have used an existing babel option.

nicolas2bert commented on .prettierrc:
```diff
@@ -1,4 +1,5 @@
 {
   "tabWidth": 2,
-  "trailingComma": "all"
```
Why do we need two different linter tools? ESLint should handle code formatting concerns.

MonPote replied:
On `metalk8s` and `xcore`. We use ESLint for code quality rules and prettier for formatting rules.
ESLint can indeed do both but I (subjective opinion) prefer prettier.

nicolas2bert replied:
I have never user Prettier.io and would love to have your opinion on "what is the advantage of using Prettier for formatting concerns over Eslint?". We could have a side discussion on this topic if you want.

nicolas2bert commented on stories/sparkline.stories.js:
```diff
@@ -29,7 +29,7 @@ const id_multi = "multiVis";
 
 export default {
-  title: "SparkLine",
+  title: "Components/Chart/SparkLine",
```
For maintainability purposes, we could assign variables for each topic/subtopic.
These values should be aligned with the `.storybook/preview.js` ones.

Cuervino replied:
Yes, agree, and it would be easier to change/to maintain.
However, as it's a limited number of items for the moment, and as it might not save time to define variables, we can keep it like that for this version.

nicolas2bert commented on stories/icons.stories.mdx:
```diff
@@ -0,0 +1,445 @@
+    <td></td>
+    <td></td>
+  </tr>
+  <tr>
```
Not sure to understand this last row.

Cuervino replied:
Agree, not clear: it's not a real icon. 
In the case of a cell with no value in a table, we use a simple "hyphen" (`&#45;`) from the font, for still displaying there is no value, without having too much visual noise.
As it's still useful to know about this specific use case, this info could be added in the "Table" component/guideline.
![image](https://user-images.githubusercontent.com/28654802/109863749-4d45ee80-7c62-11eb-979a-ca3ebbd4e989.png)

JBWatenbergScality commented on stories/guideline/button-guideline.stories.mdx:
```diff
@@ -0,0 +1,140 @@
+
+<Story id="components-button--size" />
+
+Note for updating the coreUI sizes of buttons (22FEB21)
+
+|         | current | new |
+| ------- | ------- | --- |
+| Smaller | 27      | 24  |
+| Small   | 30      | 28  |
+| default | 37      | 32  |
+| large   | 40      | 36  |
+| larger  | 48      | 40  |
```
I think maybe this should be a github issue and we should only refers the current implemented variants in the actual design system documentation

Cuervino replied:
Fully agree.

Cuervino replied:
Issue #258 opened.

JBWatenbergScality commented on stories/guideline/chart-guideline.stories.mdx:
```diff
@@ -0,0 +1,64 @@
+### Line charts
+
+- No more than 5–7 different lines
+- avoid randomly chosen color (otherwise risk if contrast issues)
```
nit: try to unify case of the first later for list items, eg: here `No` uses an upper case whereas `avoid` is lowercased

JBWatenbergScality replied:
```suggestion
- avoid randomly chosen color (otherwise risk of contrast issues)
```

JBWatenbergScality commented on stories/guideline/mdxExampleComponents.js:
```diff
@@ -0,0 +1,54 @@
+  },
+  scale: {
+    domain: ["2XX", "401", "404", "4XX", "503", "5XX"],
+    range: ["#4BE4E2", "#E45834", "#FEFA52", "#968BFF", "#BE2543", "#DC90F1"],
```
Shall we use a well defined color palette coming from the theme for charts ?

Cuervino replied:
Indeed, we do have a specific chart color palette.

Cuervino replied:
We'll need to add the chart color palette, issue #259 opened.

MonPote committed:
Update stories/guideline/chart-guideline.stories.mdx

Co-authored-by: JBWatenbergScality <75977494+JBWatenbergScality@users.noreply.github.com>

Cuervino committed:
changes for the comments

JBWatenbergScality commented on stories/Introduction.stories.mdx:
```diff
@@ -8,4 +8,25 @@ Storybook Design System is a reusable component library that helps Storybook con
+
+A Design System is an ecosystem of guidelines, tools, components, which support teams ship more efficiently consistent design.
+
+- synchronising designers, product teams, and developers
+- building a shared vocabulary for reducing communication issues
+- having one solution for one component
+- easier testing at component level
+- faster iterations with established design patterns
```
```suggestion
- Synchronising designers, product teams, and developers
- Building a shared vocabulary for reducing communication issues
- Having one solution for one component
- Easier testing at component level
- Faster iterations with established design patterns
```

JBWatenbergScality commented on stories/guideline/button-guideline.stories.mdx:
```diff
@@ -0,0 +1,138 @@
+
+Button’s label measurement:
+
+- line height: 20px
+- vertical padding: 8px
+- horizontal padding of 16px
+- space between icon & label: 8px
```
```suggestion
- Line height: 20px
- Vertical padding: 8px
- Horizontal padding of 16px
- Space between icon & label: 8px
```

JBWatenbergScality commented on stories/guideline/button-guideline.stories.mdx:
```diff
@@ -0,0 +1,138 @@
+
+## Placement
+
+-Primary buttons placed on the right of a group.  
+-Dismissive actions placed on the left. It allows the user to return to the previous screen or step in the process.
```
```suggestion
- Primary buttons placed on the right of a group.  
- Dismissive actions placed on the left. It allows the user to return to the previous screen or step in the process.
```

JBWatenbergScality commented on stories/guideline/chart-guideline.stories.mdx:
```diff
@@ -0,0 +1,64 @@
+
+Area charts
+
+- could be a single area, overlapping area, or stacked area
+- max 3-4 categories
```
```suggestion
- Could be a single area, overlapping area, or stacked area
- Max 3-4 categories
```

JBWatenbergScality commented on stories/guideline/chart-guideline.stories.mdx:
```diff
@@ -0,0 +1,64 @@
+
+### Bar and column charts
+
+- should start at a baseline of zero.
```
```suggestion
- Should start at a baseline of zero.
```

JBWatenbergScality commented on stories/guideline/button-guideline.stories.mdx:
```diff
@@ -0,0 +1,138 @@
+
+|         | px  |
+| ------- | --- |
+| Smaller | 24  |
+| Small   | 28  |
+| default | 32  |
+| large   | 36  |
+| larger  | 40  |
```
```suggestion
| Smaller | 24  |
| Small   | 28  |
| Default | 32  |
| Large   | 36  |
| Larger  | 40  |
```

JBWatenbergScality reviewed (approved):
Looks good to me 👍 , I purposed a set of suggestions to unify the case of the list items in each documentation but I let you see if it makes sense.

Cuervino committed:
Update stories/Introduction.stories.mdx

Co-authored-by: JBWatenbergScality <75977494+JBWatenbergScality@users.noreply.github.com>

Cuervino committed:
Update stories/guideline/chart-guideline.stories.mdx

Co-authored-by: JBWatenbergScality <75977494+JBWatenbergScality@users.noreply.github.com>

Cuervino committed:
Update stories/guideline/button-guideline.stories.mdx

Co-authored-by: JBWatenbergScality <75977494+JBWatenbergScality@users.noreply.github.com>

Cuervino committed:
Update stories/guideline/button-guideline.stories.mdx

Co-authored-by: JBWatenbergScality <75977494+JBWatenbergScality@users.noreply.github.com>

Cuervino committed:
Update stories/guideline/button-guideline.stories.mdx

Co-authored-by: JBWatenbergScality <75977494+JBWatenbergScality@users.noreply.github.com>

Cuervino committed:
Update stories/guideline/chart-guideline.stories.mdx

Co-authored-by: JBWatenbergScality <75977494+JBWatenbergScality@users.noreply.github.com>

Cuervino commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.0`




There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.0`




Please check the status of the associated issue None.

Goodbye cuervino.
